go/vt/vtgate/evalengine/fn_json.go

```diff
@@ -0,0 +1,366 @@
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/evalengine/internal/json"
+)
+
+type (
+	builtinJSONExtract struct {
+		CallExpr
+	}
+
+	builtinJSONUnquote struct {
+		CallExpr
+	}
+
+	builtinJSONObject struct {
+		CallExpr
+	}
+
+	builtinJSONArray struct {
+		CallExpr
+	}
+
+	builtinJSONDepth struct {
+		CallExpr
+	}
+
+	builtinJSONLength struct {
+		CallExpr
+	}
+
+	builtinJSONContainsPath struct {
+		CallExpr
+	}
+
+	builtinJSONKeys struct {
+		CallExpr
+	}
+)
+
+var _ Expr = (*builtinJSONExtract)(nil)
+var _ Expr = (*builtinJSONUnquote)(nil)
+var _ Expr = (*builtinJSONObject)(nil)
+var _ Expr = (*builtinJSONArray)(nil)
+var _ Expr = (*builtinJSONDepth)(nil)
+var _ Expr = (*builtinJSONLength)(nil)
+var _ Expr = (*builtinJSONContainsPath)(nil)
+var _ Expr = (*builtinJSONKeys)(nil)
+
+var errInvalidPathForTransform = vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "In this situation, path expressions may not contain the * and ** tokens or an array range.")
+
+func (call *builtinJSONExtract) eval(env *ExpressionEnv) (eval, error) {
+	args, err := call.args(env)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, arg := range args {
+		if arg == nil {
+			return nil, nil
+		}
+	}
+
+	doc, err := intoJSON(call.Method, args[0])
+	if err != nil {
+		return nil, err
+	}
+
+	var matches = make([]*json.Value, 0, 4)
+	var multi = len(args) > 2
+
+	for _, p := range args[1:] {
+		path, err := intoJSONPath(p)
+		if err != nil {
+			return nil, err
+		}
+
+		if path.ContainsWildcards() {
+			multi = true
+		}
+
+		path.Match(doc, true, func(v *json.Value) {
+			matches = append(matches, v)
+		})
+	}
+
+	if len(matches) == 0 {
+		return nil, nil
+	}
+	if len(matches) == 1 && !multi {
+		return matches[0], nil
+	}
+	return json.NewArray(matches), nil
+}
+
+func (call *builtinJSONExtract) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.TypeJSON, f
+}
+
+func (call *builtinJSONUnquote) eval(env *ExpressionEnv) (eval, error) {
+	arg, err := call.arg1(env)
+	if err != nil {
+		return nil, err
+	}
+	if arg == nil {
+		return nil, nil
+	}
+	j, err := intoJSON(call.Method, arg)
+	if err != nil {
+		return nil, err
+	}
+	if b, ok := j.StringBytes(); ok {
+		return newEvalRaw(sqltypes.Blob, b, collationJSON), nil
+	}
+	return newEvalRaw(sqltypes.Blob, j.MarshalTo(nil), collationJSON), nil
+}
+
+func (call *builtinJSONUnquote) typeof(env *ExpressionEnv) (sqltypes.Type, typeFlag) {
+	_, f := call.Arguments[0].typeof(env)
+	return sqltypes.Blob, f
+}
+
+func (call *builtinJSONObject) eval(env *ExpressionEnv) (eval, error) {
+	j := json.NewObject()
+	obj, _ := j.Object()
+
+	for i := 0; i < len(call.Arguments); i += 2 {
+		key, err := call.Arguments[i].eval(env)
+		if err != nil {
+			return nil, err
+		}
+		if key == nil {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "JSON documents may not contain NULL member names.")
```

Comment:
nitpick: the recommendation for error strings is to not be capitalized or end with punctuation